File: app/src/main/java/com/absinthe/kage/media/audio/RemoteAudioPlayback.java
package com.absinthe.kage.media.audio;

import android.media.session.PlaybackState;
import android.text.TextUtils;
import android.util.Log;

import com.absinthe.kage.KageApplication;
import com.absinthe.kage.connect.Const;
import com.absinthe.kage.connect.proxy.AudioProxy;
import com.absinthe.kage.device.model.AudioInfo;
import com.absinthe.kage.media.LocalMedia;
import com.absinthe.kage.media.PlayList;
import com.absinthe.kage.media.Playback;
import com.absinthe.kage.utils.NetUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class RemoteAudioPlayback implements Playback {
    private static final String TAG = RemoteAudioPlayback.class.getSimpleName();

    private Callback mCallback;
    private AudioProxy mAudioProxy = AudioProxy.getInstance();
    private PlayList mPlayList;
    private int mPlayState = PlaybackState.STATE_NONE;

    RemoteAudioPlayback() {
        mAudioProxy.setOnPlayListener(new AudioProxy.OnPlayListener() {

            @Override
            public void onCurrentPositionChanged(int duration, int position) {

            }

            @Override
            public void onPlayStateChanged(int oldState, int newState) {
                Log.d(TAG, "onPlayStateChanged: " + newState);
                if (newState == AudioProxy.PlayStatue.INVALIDATE) {
                    mPlayState = PlaybackState.STATE_NONE;
                } else if (newState == AudioProxy.PlayStatue.PLAYER_EXIT) {
                    mPlayState = PlaybackState.STATE_STOPPED;
                } else if (newState != AudioProxy.PlayStatue.DISCONNECT) {
                    switch (newState) {
                        case AudioProxy.PlayStatue.STOPPED:
                            mPlayState = PlaybackState.STATE_STOPPED;
                            break;
                        case AudioProxy.PlayStatue.TRANSITIONING:
                            mPlayState = PlaybackState.STATE_BUFFERING;
                            break;
                        case AudioProxy.PlayStatue.PLAYING:
                            mPlayState = PlaybackState.STATE_PLAYING;
                            break;
                        case AudioProxy.PlayStatue.PAUSED:
                            mPlayState = PlaybackState.STATE_PAUSED;
                            break;
                    }
                } else {
                    mPlayState = PlaybackState.STATE_STOPPED;
                }
                handlePlayState(false);
            }

            @Override
            public void onPlayIndexChanged(int index) {
                Log.d(TAG, "onPlayIndexChanged: " + index);
                mPlayList.setCurrentIndex(index);
                if (mCallback != null) {
                    mCallback.onMediaMetadataChanged(mPlayList.getCurrentMedia());
                }
            }
        });
    }

    public void playListMedia(PlayList playlist) {
        mPlayList = playlist;
        List<AudioInfo> audioInfos = new ArrayList<>();

        for (LocalMedia media : playlist.getList()) {
            if (media instanceof LocalMusic) {
                AudioInfo audioInfo = new AudioInfo();
                audioInfo.setName(media.getTitle());
                audioInfo.setUrl(media.getUrl());
                audioInfo.setArtist(((LocalMusic) media).getArtist());
                audioInfo.setAlbum(((LocalMusic) media).getAlbum());
                audioInfos.add(audioInfo);
            }
        }

        mAudioProxy.playList(mPlayList.getCurrentIndex(), audioInfos);
        mPlayState = PlaybackState.STATE_BUFFERING;

        if (mCallback != null) {
            mCallback.onMediaMetadataChanged(mPlayList.getCurrentMedia());
            mCallback.onPlaybackStateChanged(mPlayState);
        }
    }

    @Override
    public void playMedia(LocalMedia localMedia) {
        if (localMedia instanceof LocalMusic) {
            AudioInfo info = new AudioInfo();
            info.setName(localMedia.getTitle());
            info.setUrl(localMedia.getUrl());
            info.setArtist(((LocalMusic) localMedia).getArtist());
            info.setAlbum(((LocalMusic) localMedia).getAlbum());

            String ip = NetUtils.getLocalAddress();
            if (!TextUtils.isEmpty(ip)) {
                info.setCoverPath(String.format(Const.HTTP_SERVER_FORMAT, ip)
                        + KageApplication.sContext.getExternalCacheDir()
                        + File.separator + ((LocalMusic) localMedia).getAlbumId() + ".png");
            }

            mAudioProxy.play(info);
            mPlayState = PlaybackState.STATE_BUFFERING;

            if (mCallback != null) {
                mCallback.onMediaMetadataChanged(localMedia);
                mCallback.onPlaybackStateChanged(mPlayState);
            }
        }
    }

    @Override
    public void play() {
        mPlayState = PlaybackState.STATE_PLAYING;
        handlePlayState(true);
    }

    @Override
    public void pause() {
        mPlayState = PlaybackState.STATE_PAUSED;
        handlePlayState(true);
    }

    @Override
    public void seekTo(int position) {
        mAudioProxy.seekTo(position);
    }

    @Override
    public int getState() {
        return mPlayState;
    }

    @Override
    public int getDuration() {
        return mAudioProxy.getDuration();
    }

    @Override
    public int getBufferPosition() {
        return mAudioProxy.getDuration();
    }

    @Override
    public int getCurrentPosition() {
        return mAudioProxy.getCurrentPosition();
    }

    @Override
    public void setCallback(Callback callback) {
        mCallback = callback;
    }

    @Override
    public void stop(boolean fromUser) {
        mPlayState = PlaybackState.STATE_STOPPED;

        if (fromUser && mCallback != null) {
            mAudioProxy.stop();
            mCallback.onPlaybackStateChanged(mPlayState);
        }

        mAudioProxy.recycle();
    }

    private void handlePlayState(boolean fromUser) {
        if (fromUser) {
            if (mPlayState == PlaybackState.STATE_PLAYING) {
                mAudioProxy.start();
            } else if (mPlayState == PlaybackState.STATE_PAUSED) {
                mAudioProxy.pause();
            }
        }

        if (mCallback != null) {
            mCallback.onPlaybackStateChanged(mPlayState);
        }
    }
}

File: app/src/main/kotlin/com/absinthe/kage/media/audio/RemoteAudioPlayback.kt
package com.absinthe.kage.media.audio

import android.media.session.PlaybackState
import com.absinthe.kage.connect.Const
import com.absinthe.kage.connect.proxy.AudioProxy
import com.absinthe.kage.device.model.AudioInfo
import com.absinthe.kage.media.LocalMedia
import com.absinthe.kage.media.PlayList
import com.absinthe.kage.media.Playback
import com.absinthe.kage.utils.NetUtils.localAddress
import com.blankj.utilcode.util.Utils
import timber.log.Timber
import java.io.File
import java.util.*

class RemoteAudioPlayback internal constructor() : Playback {

    private var mCallback: Playback.Callback? = null
    private val mAudioProxy = AudioProxy
    private var mPlayList: PlayList? = null

    override var state = PlaybackState.STATE_NONE
        private set

    override val duration: Int
        get() = mAudioProxy.duration

    override val bufferPosition: Int
        get() = mAudioProxy.duration

    override val currentPosition: Int
        get() = mAudioProxy.currentPosition

    init {
        mAudioProxy.setOnPlayListener(object : AudioProxy.OnPlayListener {

            override fun onCurrentPositionChanged(duration: Int, position: Int) {}

            override fun onPlayStateChanged(oldState: Int, newState: Int) {
                Timber.d("onPlayStateChanged: $newState")

                when {
                    newState == AudioProxy.PlayStatue.INVALIDATE -> {
                        state = PlaybackState.STATE_NONE
                    }
                    newState == AudioProxy.PlayStatue.PLAYER_EXIT -> {
                        state = PlaybackState.STATE_STOPPED
                    }
                    newState != AudioProxy.PlayStatue.DISCONNECT -> {
                        when (newState) {
                            AudioProxy.PlayStatue.TRANSITIONING -> state = PlaybackState.STATE_BUFFERING
                            AudioProxy.PlayStatue.STOPPED -> state = PlaybackState.STATE_STOPPED
                            AudioProxy.PlayStatue.PLAYING -> state = PlaybackState.STATE_PLAYING
                            AudioProxy.PlayStatue.PAUSED -> state = PlaybackState.STATE_PAUSED
                        }
                    }
                    else -> {
                        state = PlaybackState.STATE_STOPPED
                    }
                }
                handlePlayState(false)
            }

            override fun onPlayIndexChanged(index: Int) {
                Timber.d("onPlayIndexChanged: $index")
                mPlayList?.let {
                    it.currentIndex = index
                    it.currentMedia?.let { media ->
                        mCallback?.onMediaMetadataChanged(media)
                    }
                }
            }
        })
    }

    fun playListMedia(playlist: PlayList) {
        mPlayList = playlist
        val audioInfos: MutableList<AudioInfo> = ArrayList()

        for (media in playlist.list) {
            if (media is LocalMusic) {
                audioInfos.add(AudioInfo().apply {
                    name = media.title
                    url = media.url
                    artist = media.artist
                    album = media.album
                })
            }
        }
        mAudioProxy.playList(playlist.currentIndex, audioInfos)
        state = PlaybackState.STATE_BUFFERING

        playlist.currentMedia?.let { mCallback?.onMediaMetadataChanged(it) }
        mCallback?.onPlaybackStateChanged(state)
    }

    override fun playMedia(localMedia: LocalMedia) {
        if (localMedia is LocalMusic) {
            val info = AudioInfo().apply {
                name = localMedia.title
                url = localMedia.url
                artist = localMedia.artist
                album = localMedia.album
            }

            if (localAddress.isNotEmpty()) {
                info.coverPath = (String.format(Const.HTTP_SERVER_FORMAT, localAddress)
                        + Utils.getApp().applicationContext.externalCacheDir
                        + File.separator + localMedia.albumId + ".png")
            }
            mAudioProxy.play(info)
            state = PlaybackState.STATE_BUFFERING

            mCallback?.onMediaMetadataChanged(localMedia)
            mCallback?.onPlaybackStateChanged(state)
        }
    }

    override fun play() {
        state = PlaybackState.STATE_PLAYING
        handlePlayState(true)
    }

    override fun pause() {
        state = PlaybackState.STATE_PAUSED
        handlePlayState(true)
    }

    override fun seekTo(position: Int) {
        mAudioProxy.seekTo(position)
    }

    override fun setCallback(callback: Playback.Callback) {
        mCallback = callback
    }

    override fun stop(isStop: Boolean) {
        state = PlaybackState.STATE_STOPPED
        if (isStop && mCallback != null) {
            mAudioProxy.stop()
            mCallback?.onPlaybackStateChanged(state)
        }
        mAudioProxy.recycle()
    }

    private fun handlePlayState(fromUser: Boolean) {
        if (fromUser) {
            if (state == PlaybackState.STATE_PLAYING) {
                mAudioProxy.start()
            } else if (state == PlaybackState.STATE_PAUSED) {
                mAudioProxy.pause()
            }
        }
        mCallback?.onPlaybackStateChanged(state)
    }
}